packages/dds/tree/src/feature-libraries/sequence-field/rebase.ts

```diff
@@ -222,11 +327,91 @@ function rebaseMark<TNodeChange>(
             }
             return clone(currMark);
         }
+        case "MoveOut":
+        case "ReturnFrom": {
+            if (!isSkipMark(currMark)) {
+                getOrAddEmptyToMap(moveEffects.movedMarks, baseMark.id).push(clone(currMark));
+            }
+            return 0;
+        }
         default:
             fail(`Unsupported mark type: ${baseType}`);
     }
 }
 
+function applyMoveEffects<TNodeChange>(
+    baseMarks: MarkList<TNodeChange>,
+    rebasedMarks: MarkList<TNodeChange>,
+    moveEffects: MoveEffectTable<TNodeChange>,
+    genId: IdAllocator,
+): Changeset<TNodeChange> {
+    const queue = new ComposeQueue<TNodeChange>(
+        baseMarks,
+        undefined,
+        rebasedMarks,
+        () => fail("Should not split moves while applying move effects"),
+        moveEffects,
+        false,
+    );
+    const factory = new MarkListFactory<TNodeChange>(moveEffects);
+
+    let offset = 0;
+    while (!queue.isEmpty()) {
+        const { baseMark, newMark } = queue.pop();
+        if (isObjMark(baseMark) && (baseMark.type === "MoveIn" || baseMark.type === "ReturnTo")) {
+            const movedMarks = moveEffects.movedMarks.get(baseMark.id);
+            if (movedMarks !== undefined) {
+                factory.pushOffset(offset);
+                offset = 0;
+
+                // TODO: Do moved marks ever need to be split?
+                factory.push(...movedMarks);
+                const size = movedMarks.reduce<number>(
+                    (count, mark) => count + getInputLength(mark),
+                    0,
+                );
+                factory.pushOffset(-size);
```

Comment:
Why do we need to do this?

Reply:
Suppose we have the following mark list: `[4, insert, 3, delete(2)]` and we want to move the delete to index 1. The outcome should be `[1, delete(2), 1, insert]`. Nodes 2 and 3 were previously skipped, but now they are deleted. The negative offset is a somewhat kludgy way of removing the skip marks we are replacing with the moved marks.